pkg/querier/querier.go

```diff
@@ -1148,6 +1149,62 @@ func (q *SingleTenantQuerier) DetectedFields(ctx context.Context, req *logproto.
 	}, nil
 }
 
+func (q *SingleTenantQuerier) StructuredMetadata(ctx context.Context, req *logproto.StructuredMetadataKeysRequest) (*logproto.StructuredMetadataKeysResponse, error) {
+	expr, err := syntax.ParseLogSelector(req.Query, true)
+	if err != nil {
+		return nil, err
+	}
+	params := logql.SelectLogParams{
+		QueryRequest: &logproto.QueryRequest{
+			Start:     req.Start,
+			End:       req.End,
+			Limit:     1000,
+			Direction: logproto.BACKWARD,
+			Selector:  expr.String(),
+			Plan: &plan.QueryPlan{
+				AST: expr,
+			},
+		},
+	}
+
+	iters, err := q.SelectLogs(ctx, params)
+	if err != nil {
+		return nil, err
+	}
+
+  //TODO(twhitney): we can improve this in the future as we don't need to iterate lines for structured metadata, just chunks
```

Comment:
added this as a reminder... I don't think we have good methods for getting structured metadata per chunk, but since it's stored per chunk and not per line we can improve this in the future